File: python/src/falcon_core/math/domains/labelled_domain.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.labelled_domain import LabelledDomain as _CLabelledDomain
from falcon_core.physics.device_structures.connection import Connection
from falcon_core.math.domains.domain import Domain
from falcon_core.instrument_interfaces.names.instrument_port import InstrumentPort
from falcon_core.physics.units.symbol_unit import SymbolUnit

class LabelledDomain:
    """Python wrapper for LabelledDomain."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def LabelledDomain_create_primitive_knob(cls, default_name: str, min_val: Any, max_val: Any, psuedo_name: Connection, instrument_type: str, lesser_bound_contained: Any, greater_bound_contained: Any, units: SymbolUnit, description: str) -> LabelledDomain:
        return cls(_CLabelledDomain.LabelledDomain_create_primitive_knob(default_name, min_val, max_val, psuedo_name._c, instrument_type, lesser_bound_contained, greater_bound_contained, units._c, description))

    @classmethod
    def LabelledDomain_create_primitive_meter(cls, default_name: str, min_val: Any, max_val: Any, psuedo_name: Connection, instrument_type: str, lesser_bound_contained: Any, greater_bound_contained: Any, units: SymbolUnit, description: str) -> LabelledDomain:
        return cls(_CLabelledDomain.LabelledDomain_create_primitive_meter(default_name, min_val, max_val, psuedo_name._c, instrument_type, lesser_bound_contained, greater_bound_contained, units._c, description))

    @classmethod
    def LabelledDomain_create_primitive_port(cls, default_name: str, min_val: Any, max_val: Any, psuedo_name: Connection, instrument_type: str, lesser_bound_contained: Any, greater_bound_contained: Any, units: SymbolUnit, description: str) -> LabelledDomain:
        return cls(_CLabelledDomain.LabelledDomain_create_primitive_port(default_name, min_val, max_val, psuedo_name._c, instrument_type, lesser_bound_contained, greater_bound_contained, units._c, description))

    @classmethod
    def LabelledDomain_create_from_port(cls, min_val: Any, max_val: Any, instrument_type: str, port: InstrumentPort, lesser_bound_contained: Any, greater_bound_contained: Any) -> LabelledDomain:
        return cls(_CLabelledDomain.LabelledDomain_create_from_port(min_val, max_val, instrument_type, port._c, lesser_bound_contained, greater_bound_contained))

    @classmethod
    def LabelledDomain_create_from_port_and_domain(cls, port: InstrumentPort, domain: Domain) -> LabelledDomain:
        return cls(_CLabelledDomain.LabelledDomain_create_from_port_and_domain(port._c, domain._c))

    @classmethod
    def LabelledDomain_create_from_domain(cls, domain: Domain, default_name: str, psuedo_name: Connection, instrument_type: str, units: SymbolUnit, description: str) -> LabelledDomain:
        return cls(_CLabelledDomain.LabelledDomain_create_from_domain(domain._c, default_name, psuedo_name._c, instrument_type, units._c, description))

    @classmethod
    def LabelledDomain_from_json_string(cls, json: str) -> LabelledDomain:
        return cls(_CLabelledDomain.LabelledDomain_from_json_string(json))

    def port(self, ) -> InstrumentPort:
        ret = self._c.port()
        if ret is None: return None
        return InstrumentPort._from_capi(ret)

    def domain(self, ) -> Domain:
        ret = self._c.domain()
        if ret is None: return None
        return Domain._from_capi(ret)

    def matching_port(self, port: InstrumentPort) -> None:
        ret = self._c.matching_port(port._c)
        return ret

    def lesser_bound(self, ) -> None:
        ret = self._c.lesser_bound()
        return ret

    def greater_bound(self, ) -> None:
        ret = self._c.greater_bound()
        return ret

    def lesser_bound_contained(self, ) -> None:
        ret = self._c.lesser_bound_contained()
        return ret

    def greater_bound_contained(self, ) -> None:
        ret = self._c.greater_bound_contained()
        return ret

    def in(self, value: Any) -> None:
        ret = self._c.in(value)
        return ret

    def range(self, ) -> None:
        ret = self._c.range()
        return ret

    def center(self, ) -> None:
        ret = self._c.center()
        return ret

    def intersection(self, other: LabelledDomain) -> LabelledDomain:
        ret = self._c.intersection(other._c)
        return cls._from_capi(ret)

    def union(self, other: LabelledDomain) -> LabelledDomain:
        ret = self._c.union(other._c)
        return cls._from_capi(ret)

    def is_empty(self, ) -> None:
        ret = self._c.is_empty()
        return ret

    def contains_domain(self, other: LabelledDomain) -> None:
        ret = self._c.contains_domain(other._c)
        return ret

    def shift(self, offset: Any) -> LabelledDomain:
        ret = self._c.shift(offset)
        return cls._from_capi(ret)

    def scale(self, scale: Any) -> LabelledDomain:
        ret = self._c.scale(scale)
        return cls._from_capi(ret)

    def transform(self, other: LabelledDomain, value: Any) -> None:
        ret = self._c.transform(other._c, value)
        return ret

    def equal(self, other: LabelledDomain) -> None:
        ret = self._c.equal(other._c)
        return ret

    def __eq__(self, other: LabelledDomain) -> None:
        if not hasattr(other, "_c"):
            return NotImplemented
        return self.equal(other)

    def not_equal(self, other: LabelledDomain) -> None:
        ret = self._c.not_equal(other._c)
        return ret

    def __ne__(self, other: LabelledDomain) -> None:
        if not hasattr(other, "_c"):
            return NotImplemented
        return self.not_equal(other)

    def to_json_string(self, ) -> str:
        ret = self._c.to_json_string()
        return ret

File: ocaml/src/math/domains/labelleddomain.ml
open Ctypes
open Capi_bindings
open Error_handling

(* No opens needed - using qualified names *)

class c_labelleddomain (h : unit ptr) = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.labelleddomain_destroy raw_val;
    ErrorHandling.raise_if_error ()
  ) self
end

module LabelledDomain = struct
  type t = c_labelleddomain

  let copy (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let ptr = Capi_bindings.labelleddomain_copy handle#raw in
      ErrorHandling.raise_if_error ();
      new c_labelleddomain ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.labelleddomain_from_json_string (Capi_bindings.string_wrap json) in
    ErrorHandling.raise_if_error ();
    new c_labelleddomain ptr

  let primitiveKnob (default_name : string) (min_val : float) (max_val : float) (psuedo_name : Connection.t) (instrument_type : string) (lesser_bound_contained : bool) (greater_bound_contained : bool) (units : Symbolunit.t) (description : string) : t =
    ErrorHandling.multi_read [psuedo_name; units] (fun () ->
      let ptr = Capi_bindings.labelleddomain_create_primitive_knob (Capi_bindings.string_wrap default_name) min_val max_val psuedo_name#raw (Capi_bindings.string_wrap instrument_type) lesser_bound_contained greater_bound_contained units#raw (Capi_bindings.string_wrap description) in
      ErrorHandling.raise_if_error ();
      new c_labelleddomain ptr
    )

  let primitiveMeter (default_name : string) (min_val : float) (max_val : float) (psuedo_name : Connection.t) (instrument_type : string) (lesser_bound_contained : bool) (greater_bound_contained : bool) (units : Symbolunit.t) (description : string) : t =
    ErrorHandling.multi_read [psuedo_name; units] (fun () ->
      let ptr = Capi_bindings.labelleddomain_create_primitive_meter (Capi_bindings.string_wrap default_name) min_val max_val psuedo_name#raw (Capi_bindings.string_wrap instrument_type) lesser_bound_contained greater_bound_contained units#raw (Capi_bindings.string_wrap description) in
      ErrorHandling.raise_if_error ();
      new c_labelleddomain ptr
    )

  let primitivePort (default_name : string) (min_val : float) (max_val : float) (psuedo_name : Connection.t) (instrument_type : string) (lesser_bound_contained : bool) (greater_bound_contained : bool) (units : Symbolunit.t) (description : string) : t =
    ErrorHandling.multi_read [psuedo_name; units] (fun () ->
      let ptr = Capi_bindings.labelleddomain_create_primitive_port (Capi_bindings.string_wrap default_name) min_val max_val psuedo_name#raw (Capi_bindings.string_wrap instrument_type) lesser_bound_contained greater_bound_contained units#raw (Capi_bindings.string_wrap description) in
      ErrorHandling.raise_if_error ();
      new c_labelleddomain ptr
    )

  let fromPort (min_val : float) (max_val : float) (port : Instrumentport.t) (lesser_bound_contained : bool) (greater_bound_contained : bool) : t =
    ErrorHandling.read port (fun () ->
      let ptr = Capi_bindings.labelleddomain_create_from_port min_val max_val port#raw lesser_bound_contained greater_bound_contained in
      ErrorHandling.raise_if_error ();
      new c_labelleddomain ptr
    )

  let fromPortAndDomain (port : Instrumentport.t) (domain : Domain.t) : t =
    ErrorHandling.multi_read [port; domain] (fun () ->
      let ptr = Capi_bindings.labelleddomain_create_from_port_and_domain port#raw domain#raw in
      ErrorHandling.raise_if_error ();
      new c_labelleddomain ptr
    )

  let fromDomain (domain : Domain.t) (default_name : string) (psuedo_name : Connection.t) (instrument_type : string) (units : Symbolunit.t) (description : string) : t =
    ErrorHandling.multi_read [domain; psuedo_name; units] (fun () ->
      let ptr = Capi_bindings.labelleddomain_create_from_domain domain#raw (Capi_bindings.string_wrap default_name) psuedo_name#raw (Capi_bindings.string_wrap instrument_type) units#raw (Capi_bindings.string_wrap description) in
      ErrorHandling.raise_if_error ();
      new c_labelleddomain ptr
    )

  let equal (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.labelleddomain_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.labelleddomain_not_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.labelleddomain_to_json_string handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let port (handle : t) : Instrumentport.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.labelleddomain_port handle#raw in
      ErrorHandling.raise_if_error ();
      new c_instrumentport result
    )

  let domain (handle : t) : Domain.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.labelleddomain_domain handle#raw in
      ErrorHandling.raise_if_error ();
      new c_domain result
    )

  let matchingPort (handle : t) (port : Instrumentport.t) : bool =
    ErrorHandling.multi_read [handle; port] (fun () ->
      let result = Capi_bindings.labelleddomain_matching_port handle#raw port#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let lesserBound (handle : t) : float =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.labelleddomain_lesser_bound handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let greaterBound (handle : t) : float =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.labelleddomain_greater_bound handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let lesserBoundContained (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.labelleddomain_lesser_bound_contained handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let greaterBoundContained (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.labelleddomain_greater_bound_contained handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let in_ (handle : t) (value : float) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.labelleddomain_in handle#raw value in
      ErrorHandling.raise_if_error ();
      result
    )

  let range (handle : t) : float =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.labelleddomain_range handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let center (handle : t) : float =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.labelleddomain_center handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let intersection (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.labelleddomain_intersection handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_labelleddomain result
    )

  let union (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.labelleddomain_union handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_labelleddomain result
    )

  let isEmpty (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.labelleddomain_is_empty handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let containsDomain (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.labelleddomain_contains_domain handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let shift (handle : t) (offset : float) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.labelleddomain_shift handle#raw offset in
      ErrorHandling.raise_if_error ();
      new c_labelleddomain result
    )

  let scale (handle : t) (scale : float) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.labelleddomain_scale handle#raw scale in
      ErrorHandling.raise_if_error ();
      new c_labelleddomain result
    )

  let transform (handle : t) (other : t) (value : float) : float =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.labelleddomain_transform handle#raw other#raw value in
      ErrorHandling.raise_if_error ();
      result
    )

end